File: TrainingCell/Source/Checkers/AgentPack.cpp
//Copyright (c) 2023 Denys Dragunov, dragunovdenis@gmail.com
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files(the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and /or sell
//copies of the Software, and to permit persons to whom the Software is furnished
//to do so, subject to the following conditions :

//The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include "../../Headers/Checkers/AgentPack.h"
#include "../../Headers/Checkers/RandomAgent.h"
#include "../../../DeepLearning/DeepLearning/MsgPackUtils.h"

namespace TrainingCell::Checkers
{
	AgentPack::AgentPack(const AgentTypeId id, std::unique_ptr<Agent>&& agent_ptr) :
	_agent_id(id), _agent_ptr(std::move(agent_ptr)) {}

	AgentPack::AgentPack(const AgentPack& anotherPack): _agent_id(anotherPack._agent_id)
	{
		if (_agent_id == TdLambdaAgent::TYPE_ID())
		{
			_agent_ptr = std::make_unique<TdLambdaAgent>(dynamic_cast<const TdLambdaAgent&>(anotherPack.agent()));
			return;
		}
		if (_agent_id == TdlEnsembleAgent::TYPE_ID())
		{
			_agent_ptr = std::make_unique<TdlEnsembleAgent>(dynamic_cast<const TdlEnsembleAgent&>(anotherPack.agent()));
			return;
		}
		if (_agent_id == RandomAgent::TYPE_ID())
		{
			_agent_ptr = std::make_unique<RandomAgent>(dynamic_cast<const RandomAgent&>(anotherPack.agent()));
			return;
		}

		throw std::exception("Not implemented");
	}

	void AgentPack::msgpack_unpack(msgpack::object const& msgpack_o)
	{
		//Read identifier only
		msgpack::type::make_define_array(_agent_id).msgpack_unpack(msgpack_o);
		//Instantiate an agent based on the identifier
		if (_agent_id == TdLambdaAgent::TYPE_ID())
		{
			auto proxy = TdLambdaAgent();
			//Read once again, but this time we read the instance of the agent as well
			msgpack::type::make_define_array(_agent_id, proxy).msgpack_unpack(msgpack_o);
			_agent_ptr = std::make_unique<decltype(proxy)>(std::move(proxy));
			return;
		}
		if (_agent_id == TdlEnsembleAgent::TYPE_ID())
		{
			auto proxy = TdlEnsembleAgent();
			//Read once again, but this time we read the instance of the agent as well
			msgpack::type::make_define_array(_agent_id, proxy).msgpack_unpack(msgpack_o);
			_agent_ptr = std::make_unique<decltype(proxy)>(std::move(proxy));
			return;
		}

		throw std::exception("Not implemented");
	}

	Agent& AgentPack::agent()
	{
		if (_agent_ptr == nullptr)
			throw std::exception("Layer is not initialized");

		return *_agent_ptr;
	}

	const Agent& AgentPack::agent() const
	{
		if (_agent_ptr == nullptr)
			throw std::exception("Layer is not initialized");

		return *_agent_ptr;
	}

	void AgentPack::save_to_file(const std::filesystem::path& file_path) const
	{
		DeepLearning::MsgPack::save_to_file(*this, file_path);
	}

	AgentPack AgentPack::load_from_file(const std::filesystem::path& file_path)
	{
		return DeepLearning::MsgPack::load_from_file<AgentPack>(file_path);
	}

	bool AgentPack::operator == (const AgentPack& anotherPack) const
	{
		return _agent_id == anotherPack._agent_id &&
			((_agent_ptr == nullptr && anotherPack._agent_ptr == nullptr) ||
				agent().equal(anotherPack.agent()));
	}

	bool AgentPack::operator != (const AgentPack& anotherPack) const
	{
		return !(*this == anotherPack);
	}
}

File: Monitor/Checkers/AgentPack.cs
//Copyright (c) 2023 Denys Dragunov, dragunovdenis@gmail.com
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files(the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and /or sell
//copies of the Software, and to permit persons to whom the Software is furnished
//to do so, subject to the following conditions :

//The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

using System;

namespace Monitor.Checkers
{
    /// <summary>
    /// Wrapper class for the naive AgentPack structure
    /// </summary>
    class AgentPack : Agent
    {
        /// <summary>
        /// Pointer to an instance of native object
        /// </summary>
        private IntPtr _packPtr;

        /// <summary>
        /// Constructs by initializing native resources from file
        /// </summary>
        public AgentPack(string filePath)
        {
            _packPtr = DllWrapper.AgentPackLoadFromFile(filePath);

            if (_packPtr == IntPtr.Zero)
                throw new Exception("Failed to load from file");
        }

        /// <summary>
        /// Packs the given agent
        /// </summary>
        public AgentPack(TdLambdaAgent agent)
        {
            if (agent == null)
                throw new Exception("Invalid input");

            _packPtr = DllWrapper.PackTdLambdaAgent(agent.Ptr);
        }

        public AgentPack(EnsembleAgent agent)
        {
            if (agent == null)
                throw new Exception("Invalid input");

            _packPtr = DllWrapper.PackTdlEnsembleAgent(agent.Ptr);
        }

        /// <summary>
        /// Releases native resources
        /// </summary>
        public override void Dispose()
        {
            if (_packPtr == IntPtr.Zero)
                return;

            if (!DllWrapper.AgentPackFree(_packPtr))
                throw new Exception("Failed to release agent pointer");

            _packPtr = IntPtr.Zero;
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Saves native resource to the given file on disk
        /// </summary>
        public void SaveToFile(string filePath)
        {
            if (!DllWrapper.AgentPackSaveToFile(_packPtr, filePath))
                throw new Exception("Saving failed");
        }

        /// <summary>
        /// Read-only access to the packed agent
        /// </summary>
        public override IntPtr Ptr => DllWrapper.AgentPackGetAgentPtr(_packPtr);

        /// <summary>
        /// Training mode parameter of the packed agent
        /// </summary>
        public bool TrainingMode
        {
            get => DllWrapper.AgentGetTrainingMode(Ptr).ToBool();

            set
            {
                if (!DllWrapper.AgentSetTrainingMode(Ptr, value))
                    throw new Exception("Failed to set parameter");
            }
        }

        /// <summary>
        /// Can train flag of the packed agent
        /// </summary>
        public bool CanTrain => DllWrapper.AgentGetCanTrainFlag(Ptr).ToBool();
    }
}
